src/network/core/http_winhttp.cpp

```diff
@@ -0,0 +1,301 @@
+		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), buffer, sizeof(buffer), NULL) == 0) {
+		return fmt::format("unknown error {}", error_code);
+	}
+
+	return buffer;
+}
+
+/**
+ * Callback from the WinHTTP library, called when-ever something changes about the HTTP request status.
+ *
+ * The callback needs to call some WinHttp functions for certain states, so WinHttp continues
+ * to read the request. This also allows us to abort when things go wrong, by simply not calling
+ * those functions.
+ * Comments with "Next step:" mark where WinHttp needs a call to continue.
+ *
+ * @param code    The code of the event.
+ * @param info    The information about the event.
+ * @param length  The length of the information.
+ */
+void NetworkHTTPRequest::WinHttpCallback(DWORD code, void *info, DWORD length)
+{
+	switch (code) {
+		case WINHTTP_CALLBACK_STATUS_RESOLVING_NAME:
+		case WINHTTP_CALLBACK_STATUS_NAME_RESOLVED:
+		case WINHTTP_CALLBACK_STATUS_CONNECTING_TO_SERVER:
+		case WINHTTP_CALLBACK_STATUS_CONNECTED_TO_SERVER:
+		case WINHTTP_CALLBACK_STATUS_SENDING_REQUEST:
+		case WINHTTP_CALLBACK_STATUS_REQUEST_SENT:
+		case WINHTTP_CALLBACK_STATUS_RECEIVING_RESPONSE:
+		case WINHTTP_CALLBACK_STATUS_RESPONSE_RECEIVED:
+		case WINHTTP_CALLBACK_STATUS_CLOSING_CONNECTION:
+		case WINHTTP_CALLBACK_STATUS_CONNECTION_CLOSED:
+		case WINHTTP_CALLBACK_STATUS_HANDLE_CREATED:
+		case WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING:
+			/* We don't care about these events, and explicitly ignore them. */
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_REDIRECT:
+			/* Make sure we are not in a redirect loop. */
+			if (this->depth++ > 5) {
+				Debug(net, 0, "HTTP request failed: too many redirects");
+				this->callback->OnFailure();
+				this->finished = true;
+				return;
+			}
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE:
+			/* Next step: read response. */
+			WinHttpReceiveResponse(this->request, nullptr);
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_HEADERS_AVAILABLE:
+		{
+			/* Retrieve the status code. */
+			DWORD status_code = 0;
+			DWORD status_code_size = sizeof(status_code);
+			WinHttpQueryHeaders(this->request, WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER, WINHTTP_HEADER_NAME_BY_INDEX, &status_code, &status_code_size, WINHTTP_NO_HEADER_INDEX);
+			Debug(net, 3, "HTTP request status code: {}", status_code);
+
+			/* If there is any error, we simply abort the request. */
+			if (status_code >= 400) {
+				Debug(net, 0, "HTTP request failed: status-code {}", status_code);
+				this->callback->OnFailure();
+				this->finished = true;
+				return;
+			}
+
+			/* Next step: query for any data. */
+			WinHttpQueryDataAvailable(this->request, nullptr);
+		} break;
+
+		case WINHTTP_CALLBACK_STATUS_DATA_AVAILABLE:
+		{
+			/* Retrieve the amount of data available to process. */
+			DWORD size = *(DWORD *)info;
```

Comment:
Should we do some sort of sanity check on this size?

Reply:
What sanity check do you suggest? I can't really think of one ... that would mean the Windows API did something weird, so I find it difficult to think about what to check for.

Reply:
I don't really know either, though that could mean allocating a few GB of memory.
I also have no clue what happens when reading/allocating a buffer smaller that this size. Would that just call this callback multiple times? Then limiting to something like 1 MB might be useful to not use a lot of memory for a 32 bpp base set.

Reply:
I did some checking, also with other projects. Both the Windows manual suggests to do it like this, but software like `wine` and `duckstation` also do it like this. So I am guessing, but I cannot confirm that, that Windows take care of this. Otherwise a lot of these applications would constantly run out of memory :)

WinHttp is just a bit weird to understand, so I rather stick with what the manual says, as that is most likely to work. If we ever get reports about out-of-memory issues, we can pick it up from there :)